### PR TITLE
feat: add compilation-time-defined variable type called DEPLOY

### PR DESCRIPTION
**Pull request by ayrc**
# Compilation-time-defined variable type called DEPLOY

Add a new type of variable called DEPLOY variable, which allows for compile-time configuration for specific deployment targets.

## Description

Adds/modifies configuration and syntax.

* extends Litexa configuration

* extends `when` statements with postfix conditionals and limited-scope file exclusion statement

* `litexa test` and `litexa model` are now deployment target specific

* skill configuration allows deployment target overrides

## Motivation and Context

It allows changing the language model programmatically for each deployment version, which may help with testing. It also allows similar skills (where most logic and interaction is the same) to share a base Litexa project.

## Testing

Added litexa test to test package for DEPLOY variables and file exclusion statement.
Added manifest.spec.coffee.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/alexa-games/litexa/issues
[license]: https://www.apache.org/licenses/LICENSE-2.0
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: docs/book/deployment.md [hunk 1]
```diff
@@ -451,13 +469,13 @@ Once again, you can view the result skill invocation names in the deployed model
...
+## Extra Note: Including `production` in the Deployment Target Name
 
-There is a special deployment target name called `production`.
-For deployment targets that are *not* this, a `\ (development)` will be
-automatically appended to the name of your skill, with `development`
-specified as your named deployment target. We recommend creating and
-using this deployment target for your live skill.
+There is a special condition on deployment target names. If the name does *not*
+have the word `production` in it, a ` (development)` will be automatically
+appended to the name of your skill, with `development` specified as your named
```
Comment (joseifert) [inline docs/book/deployment.md]: Would reword to more clearly indicate that the deployment target name (in this case 'development') is what's appended in parentheses.

File: docs/book/expressions.md [hunk 2]
```diff
@@ -227,6 +227,60 @@ askName
...
+      DEPLOY: {
+        mode: 'challenge'
+      }
+    },
+    ...
+```
+
+These variables will then be available for use in your Litexa code with their
+values dependent on which deployment target you specified in the litexa command
+(`litexa test`, `litexa deploy`, etc.).
```
Comment (joseifert) [inline docs/book/expressions.md]: Might be worth adding a warning to specify that the `DEPLOY` object should only be used for properties, and not for methods (which would presumably just be trimmed by the JSON stringification).

File: packages/litexa/src/parser/litexa.pegjs [hunk 3]
```diff
@@ -370,6 +373,54 @@ DefineStatement "database type definition"
...
+/* litexa [exclude file]
+The `exclude file` statement lets you specify a static [expression](
+#expression) to exclude all the contents of the file. You might want to exclude
+a file of test states from a production build, or exclude a file from versions
+of your skill, as defined by your [DEPLOY](#deploy-variable) variables.
+
+```coffeescript
+exclude file unless DEPLOY.DEBUG
+# excludes a file's contents unless the DEPLOY variable DEBUG was set
+```
```
Comment (joseifert) [inline packages/litexa/src/parser/litexa.pegjs]: Might be a little confusing since "tests" in the reader's understanding would be excluded regardless. Could perhaps reword along the lines of:

The `exclude file` statement can be used to skip a `.test.litexa` file if the tests within aren't relevant and would otherwise fail for the deployment target being tested. For instance:
```coffeescript
# campaign.test.litexa
exclude file unless DEPLOY.ENABLE_CAMPAIGN
# ... tests for the campaign mode that assume it's enabled
```

File: packages/litexa/src/parser/litexa.pegjs [hunk 4]
```diff
@@ -370,6 +373,54 @@ DefineStatement "database type definition"
...
+
+Note: this statement only works on files that exclusively have tests or states
+that don't have any transitions to them.
```
Comment (joseifert) [inline packages/litexa/src/parser/litexa.pegjs]: a couple trivial notes:
1. Upper case "this".
2. Think you could wrap this in a `:::warning`
3. Not sure if it's even worth noting this, since the non-test files this would realistically apply to is probably zero to none?

File: docs/book/expressions.md [hunk 5]
```diff
@@ -227,6 +227,60 @@ askName
...
+
+```javascript
+const deploymentConfiguration = {
+  name: 'cats-in-space',
+  deployments: {
+    development: {
+      ... // other configuration
+      DEPLOY: { // your DEPLOY variables go in this object
+        debug: true,
+        mode: 'practice'
```
Comment (joseifert) [inline docs/book/expressions.md]: Would recommend casing these deploy properties consistently (lower case here, fully upper case in file exclusion example). Personally favor the full upper case since these should typically treated as (static) constants, but either way's fine.

File: docs/book/deployment.md [hunk 6]
```diff
@@ -353,10 +353,28 @@ Litexa takes your `skill.coffee/js/ts` (we might refer to this as `skill.*` in d
 to create your skill manifest file. You can find the constructed skill manifest file after a deployment
 in `.deploy/{yourDeploymentTarget}/skill.json`.
 
-If you use any litexa extensions in your project, they may add the APIs they use to your skill manifest automatically.
+If you use any Litexa extensions in your project, they may add the APIs they use to your skill manifest automatically.
 Otherwise, you will need to add the required fields to your `skill.*`. Likewise, if you use an API Litexa does not
 explicitly support (e.g. Gadget Controller), you will need to add it to your `skill.*`.
 
+#### Deployment Target Overrides
```
Comment (joseifert) [inline docs/book/deployment.md]: Should at the very least mention `DEPLOY` variables and link to them from this chapter.

File: docs/book/expressions.md [hunk 7]
```diff
@@ -227,6 +227,60 @@ askName
     @name = $name
 ```
 
+## DEPLOY Variables
+
+DEPLOY variables are a type of memory storage variable that you can define as
+part of your Litexa configuration to exist at compilation time. They can be
+used like the other types of variables, but can also be used in [`when`
+statements](/reference/#when) and [file exclusion statements](/reference/#exclude-file).
```
[the remark this answers is not in the document]
Comment (ayrc) [reply docs/book/expressions.md]: In the introduction of this chapter, it states 2 variable types without explaining them much. I'll add an explanation on those and reword with your suggestions. I do kind of prefer to leave the in depth explanations and examples to their specific sections (in the syntax) though.

File: packages/litexa/src/parser/state.coffee [hunk 8]
```diff
@@ -300,7 +300,12 @@ class lib.State
       try
         model = intent.toModelV2(context)
       catch err
-        throw "failed to write language model for state `#{@name}`: #{err}"
+        if err.location
+          throw err
```
[the remark this answers is not in the document]
Comment (ayrc) [reply packages/litexa/src/parser/state.coffee]: It looks like `formatLocationStart` is used for formatting the location in an error message, so it's not used here. Digging into intent.toModelV2(), this is where the static conditional check can throw a ParserError, in which case we want to bubble that up instead of the more generic error. I've added a comment explaining this in the second revision.

File: docs/book/deployment.md [hunk 9]
```diff
@@ -465,17 +467,23 @@ Combining the above examples will net you `cloudy cats dev` in en-GB and
...
 ## Extra Note: Including `production` in the Deployment Target Name
 
 There is a special condition on deployment target names. If the name does *not*
-have the word `production` in it, a ` (development)` will be automatically
-appended to the name of your skill, with `development` specified as your named
-deployment target. We recommend labelling only your production skill deployment
-target(s) with `production`.
+have the word `production` in it, a ` (<target>)` will be automatically
+appended to the name of your skill, with `<target>` being the name of the
+deployment target you used to deploy the skill. We recommend labelling only
```
Comment (joseifert) [inline docs/book/deployment.md]: trivial: "labelling" is Br. whereas "labeling" is Am. - think we've pretty consistently used Am. spelling throughout the docs.
